### DISCUSSION TITLE
How do I get the Prove transaction hash, given a withdraw initiated in the L2 ?

### DISCUSSION THREAD

**Discussion by gndelia** [ethereum-optimism/developers#609] [Application Development]
### Did you check the documentation?

- [X] Yes

### Did you check for duplicate questions?

- [X] Yes

### Issue Description

I had this question created as an issue https://github.com/ethereum-optimism/optimism/issues/9974
I'm using the OP SDK (I know it's deprecated),  and given a completed withdrawal transaction hash, I know I can use [getMessageReceipt](https://sdk.optimism.io/classes/crosschainmessenger#getMessageReceipt) to get the Claim transaction hash.

```ts
const l1Rcpt = await crossChainMessenger.getMessageReceipt(l2TxHash)
const l1TxHash = l1Rcpt.transactionReceipt.transactionHash
```

How can I get the prove transaction hash? I can't find a method in the docs that returns so. Should I do a mix of different calls?

Thanks

### Additional Information

I'm using  @eth-optimism/sdk v3.2.1

**Comment by joohhnnn** [comment]
Hi there if the `prove transaction hash` you mentioned here is fault proof distribute game hash, you can find it by the specify game, and you can find every games by game factory. you can refer this [diagram](https://mermaid.live/view#pako:eNq9WNtu2zgQ_RWB--ICsmU5V6mLBdIkzRZIN0XSosDCL7Q4tolIopakkrhJ_n1Hki-kRCfNYt08BLzMmTkczoXyI0kEAxKTmaTF3Lu8Huce_qly0iyMyfmDBpnT1DsVuZY00WpMGqHqb7V7fn06GvbWsvV0g3jnBByNQguB8w4EclYPOrQuQ-8mo1K7aV2GN5rmjEr2QXI2g1574Z0p29heS5pTS-5UCqXOREZ5_hmUgnwGsudeNnBXheYZV9kXITVNe_bUkLspC5DJHLXgkaZ81msvmLILpSFbyRkTQ-aMq6LUcEEz-Ij-EXLR6y4Z8h9pmWpDotdeMHVDShfAvp9__bNnjFuX1r21bwqk9ynHO0fzXOTWpVWbqlf_f1VRc-atqi5KvGlO897Ft5Prs08nf7kVJilPbjvBgurnWhcqDoIZ1_NyMkhEFoCeg4Qy64vlBQbrgZYAAYM7SEURFDS5pTNQQbIKzP4EmBTJbaBkElyGQdveQIl0TOqYdm2M87HeUDWDc_dETWsGze6yTdKVEbsn67JqkN6-bZEfXZUaI_4KjaS797BlbUPWsWyStGvIrkna1tYkXcsmyXb52jXNtr01UfeGRdWooDunadjaUOwsmgWqW7Z3SpI15oKu2TXfbVsm63b7-CWc20bXjN0bVRCM9dLLmz72S6jeo77AMLpxbnsNWeadx5b3e7__x1PGcx1MSpkjBZqrKcinTj_rgKuH1gtos7qb3bLTJ2sdWEhZICEBfgdehoUVj_60pQes9FhN7D9rcXaZWtvyBgK3VrtwbTlbf4B6_ikBc62gpQJPaaoR3C4mziO9Few8yduUtFrC_wVOaqknq0K-hplVyU4Zk3iUJ0fx-in8km87bRusoyBW155IQJDyeI7oPMGRmG5T0alOdRSiKsCA0ZJjLnsTkbPqBJt8NNOhfqg28cagEIrr4B6LBZP03kPRoM7Sbdn4IrgOJXcmNrg6xqW4gwAeIMETeCswerJO5eWb2B3sGyWNx7xlSao9_8KVGcbxy4snvKjAPG_j3R5vau3qZV6rmaSYqClXuq2hzbqDxBDGJPEwxApINLAmaPSicOX3qitRTDKYevCg1x-NGGdS3EL823Qa-c24f8-Znsej4uF9C5iGXdxkMv0J3PnViYGYvIrAjJX4bbNhF9FXMapO0foL6y2wRHK8SoybFWg4Hb4IstuQ3-orpnMtVDsPfDu8fXcV9NsFy7feasaNWMaaSF160dr5QHWCbdz_IgWmHQ7WYbXxnwWwo8nv5obfDnfffEms3Pue-CQDiYdjJCaPlYUxwQcFZgeJcchgWqmpmv0zitJSi5tFnpBYyxJ8IkU5m5N4SlOFs7JgmHlnnOLncLYSKWj-txDZWgjnJH4kDyQOw9Eg2js6CKP9cHR0fDg88smCxP0wHERRNDreGw4PD6KDcH_07JMftY5wMDyO9vejo8O9UXQ83NtHCDCO5_3c_F5U_2z0_C-RhmRh).

**Reply by gndelia** [reply]
Hey @joohhnnn ! Thanks for answering. I am a bit of a newbie on the whole OP flow, so I am sorry if what follows is not correct 😅 .

I am not doing this on Optimism but on a Forked L2 from OP.

The prove step I am referring to is [this one](https://docs.optimism.io/stack/protocol/rollup/withdrawal-flow#withdrawal-proving-transaction), after calling `proveMessage`.

I think this forked L2 is not using dispute games (Our L2 is not using FPAC  contracts yet)

**Reply by joohhnnn** [reply]
oh I think I got your question 
PLZ check this [diagram](https://mermaid.live/view#pako:eNq1V99v4jgQ_lci3wsrBUJSWprs6aTdLto9iT2qcvdy4sUkA1iQOGc7Ld3S__2GhAQ7CWV7OvIQxfZ8M5_nl50XEvIISECWgqYra_wwSyx8ZDYvJmZktFUgErqx7niiBA2VnJFCaP-Uq6OHO6_fqWTz4RHxoRUw9FwDgeMGBJKo-GjwGrvWNKZCtfMau1NFk4iK6LNg0RI69YkPumxhvJLUh4bcneBSfuExZcl3kBKSJYhO-7SO8yaZSjM1QZYb1G8MNblJqljMZHzPhaKbjjnU5KZZCiJcoTXc-oItO_UJXfZZKohLOW1w1r9_SRDW7wmGB33LeGK4d78oO_n7rKLC6klV94KnHBV17h8m95Pp6EFj_zXDeDGadO6-fRqPR398rVZr5sINC9eNoKPxlVKpDBxnydQqm_dCHjugViAgi7v84GCn-lACwIngETY8dVIarukSpBOWCdadQyR4uHakCJ2x69Tt9STfzEiem20Ls2SmjlT1JLs8Ud2aRrM5bZJsy-zLk22zqpE-vWyQN8rs4qQNa0eyLdM6SbPGL03StFaRbJvWSdbby6Vp1u1VRNsXDKpah7s4Tc3WkWJjUm9TZa-zut3fdmkxsKSiCiye54nc1fK2wJV9MMcJiPkjWCx5pBsW_Qy8eBtHtfXrXlXMEuXMM5Hg3mkiFyB2jSbaAO-P6TfQekvRzTeac64DqzdyBITAcEsxVjP6fHei8ZR6jM75n7W0trZc2yH0TrtWs1pO7K3bQz3_ZCCerZRmZZB3jVJq3dJ7wa07eZ-SWh_6v8BhLrUzyvKswaIsDon9Zkbn948iZhEgiCnnCSs9EvTJGv35zckz_VRGvwnOw9GezQXuUMCP4MAWwgwrsATjbvJyOFx12hOmvLVQDBwsLNiq6haL-xZ8DcEvi4VvF9_dJxapVeCl24814MZt4ubzxU_gRpNPGmJ-FoFZIPAKd2Tn07MYmYc9v0i-BxYKpliIfixB_UX_TZDZ2uxar9Kda6DqeWGb4bbbK8uuF4FtHDpaRAxjRdocvGisfKYqxDPJLg8Hu-r2R_8ZADOh7NpFp_TeR2KTGARyj_D37mWvYEbw8IthRgL8xH2v96fTK8rRTPHpcxKSQIkMbCJ4tlyRYEE3EkdZGmElfmEUb_RxKZLS5G_O40oIxyR4IVsSuK7X86-G164_cL3h7U1_aJNnEnRdt-f7vnd71e_fXPvX7sB7tcmPXIfb69_6g4E_vLny_Nv-1QAhEDHFxffi7zT_SX39F-0fEQw)

If I understand correctly you need user's withdraw Application transaction hash, you can make a filter with user address and L1StandardBridge address. If not, you can refer this way to get any data you wanted. First, you need to get a clear understanding of how the components interact with each other. The diagram above is a great starting point.

**Reply by gndelia** [reply]
Sorry for the delay, I've been working on other things 😓  Perhaps you can still help me

@joohhnnn  what I need is something like this:

This is a withdrawal initiated in an L2 [0xdbd2f1d46a2f508a6db03e7e5fcfd8a482c1461cc5b3db5a99311fd5b9b55469](https://testnet.explorer.hemi.xyz/tx/0xdbd2f1d46a2f508a6db03e7e5fcfd8a482c1461cc5b3db5a99311fd5b9b55469) (for all purposes, it's a fork from Optimism)
Now, I've just proved that withdrawal. Here in Sepolia [0xa2312f662244db3aa8c09251da5dd396e9c03784a865b21768d65b6e699f9806](https://sepolia.etherscan.io/tx/0xa2312f662244db3aa8c09251da5dd396e9c03784a865b21768d65b6e699f9806) is the transaction hash.


Now, let's assume I have a page that wants to show this information from scratch (so the 3 transactions are shown in some kind of flow graph). I have a subgraph that, given the user wallet on the L2, gets me the first transaction hash. So I have `0xdbd2f1d46a2f508a6db03e7e5fcfd8a482c1461cc5b3db5a99311fd5b9b55469` ( the withdrawal initiated in the L2) . What I don't know how to do is to get from there to the Prove transaction hash (`0xa2312f662244db3aa8c09251da5dd396e9c03784a865b21768d65b6e699f9806`).

I just tried creating a subgraph on the L2 bridge, and monitoring the `WithdrawalProven` events, only to find out that the `withdrawalHash` that's emitted is a hash of the Withdrawal object, and not the TX the user itself got. 

This fork is not using Dispute games yet.

Can you help me getting that information? I'm a bit lost. 

Thanks 🙏🏽 


edit: In addition to that, I also need to get the claim transaction. I just claimed (after writing that) that transaction in Sepolia, and I got the has [0x3aac9720293c8521190cc274b184ddf1ba301e616d4cf9c6fe853e88ae31296c](https://sepolia.etherscan.io/tx/0x3aac9720293c8521190cc274b184ddf1ba301e616d4cf9c6fe853e88ae31296c). The idea would be to also get this hash starting from the withdrawal hash from the first transaction

edit2:
I was able to generate that `withdrawalHash` by doing this (with viem)
```ts
const encoded = encodeAbiParameters(
    parseAbiParameters('uint256, address, address, uint256, uint256, bytes'),
    [
      nonce,
      sender,
      target, // transaction.target,
      value,
      gasLimit,
      data,
    ],
  )
const withdrawalHash = keccak256(encoded)
```

So I guess I could still use my subgraph that's filtering events, and match by that 🤔
